sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExpressionCodegen.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.codegen
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions._
+
+/**
+ * Defines APIs used in expression code generation.
+ */
+object ExpressionCodegen {
+
+  /**
+   * Given an expression, returns the all necessary parameters to evaluate it, so the generated
+   * code of this expression can be split in a function.
+   * The 1st string in returned tuple is the parameter strings used to call the function.
+   * The 2nd string in returned tuple is the parameter strings used to declare the function.
+   *
+   * Returns `None` if it can't produce valid parameters.
+   *
+   * Params to include:
+   * 1. Evaluated columns referred by this, children or deferred expressions.
```

Comment:
I think the whole stage codegen framework guarantees that, if one of the `ctx.currentVars` is referenced, it must be evaluated. Did you hit cases that break it?

Reply:
You mean deferred expressions? Yes, we can't pass tests if we don't consider it. Some variables can be deferred to evaluate in the operators.

Reply:
In whole stage codegen framework, each operator defines the input attributes it needs, and the framework will pre-evaluate them. This means, if one of the `ctx.currentVars` is referred, it must have already been evaluated.

Do you still remember which tests were failed? Maybe it's a bug.

Reply:
For example, `ProjectExec` doesn't pre-evaluate all input attributes from its child operator.

Reply:
I just did a quick test. Some tests like `org.apache.spark.ml.tuning.TrainValidationSplitSuite` and `org.apache.spark.ml.classification.GBTClassifierSuite` will fail.

Reply:
yea, because the expressions in `ProjectExec` may only need a subset of the input attributes. For those input attributes that are not pre-evaluated, they should not be referred, or do I miss something?

Reply:
Currently `ProjectExec` only pre-evaluates the input attributes which are referred more than twice. Otherwise, we defer the evaluation until input attribute is actually used.

Reply:
`FilterExec` is another example. It doesn't pre-evaluate any input attributes and defer the evaluation to when we evaluate the related predicates.

Reply:
ah i see, thanks for the explanation!